spec/consensus/gossip/crdt.md

```diff
@@ -130,48 +129,178 @@ Consider the following definition for **Eventual Convergence**.
 We pointed out [previously](#on-the-need-for-the-gossip-communication-property) that the Gossip Communication property is overkill for Tendermint because it requires even stale messages to be delivered.
-Removing tuple is exactly how stale messages get removed.
+We remove entries corresponding to stale messages and never deliver them them.
 
 [^proof]:  TODO: do we need to extend here?
 
 ### When to remove
 
 > **TODO** Define conditions for tuple removal. Reference
 
+## The tuple space as a CRDT
+
+Conflict-free Replicated Data Types (CRDT) are distributed data structures that explore commutativity in update operations to achieve [Strong Eventual Consistency](https://en.wikipedia.org/wiki/Eventual_consistency#Strong_eventual_consistency).
+As an example of CRDT, consider a counter updated by increment operations, known as Grown only counter (G-Counter): as long as the same set of operations are executed by two replicas, their views of the counter will be the same, irrespective of the execution order.
+
+More relevant CRDT are the Grow only Set (G-Set), in which operations add elements to a set, and the [2-Phase Set](https://en.wikipedia.org/wiki/Conflict-free_replicated_data_type#2P-Set_(Two-Phase_Set)) (2P-Set), which combines two G-Set to collect inclusions and exclusions to a set.
+
+CRDT may be defined in two ways, operation- and state-based.
+Operation-based CRDT use reliable communication to ensure that all updates (operations) are delivered to all replicas.
+If the reliable communication primitive precludes duplication, then applying all operations will lead to the same state, irrespective of the delivery order since operations are commutative.
+If duplications are allowed, then the operations must be made idempotent somehow.
+
+State-based CRDT do not rely on reliable communication.
+Instead they assume that replicas will compare their states and converge two-by-two using a merge function;
+as long as the function is commutative, associative and idempotent, the states will converge.
+For example, in the G-Set case, the merge operator is simply the union of the sets.
+
+The two approaches for converging the message sets in the Tendermint algorithm described [earlier](#nodes-state-as-a-tuple-space), without the deletion of entries, correspond to the operation- and state-based [Grow-only Set](https://en.wikipedia.org/wiki/Conflict-free_replicated_data_type#G-Set_(Grow-only_Set)) CRDT;
+if removals must be handled, then using a 2P-Set is an option.
+To the best of our knowledge, no existing CRDT supports superseding of elements, until now.
+
+### About supersession
+
+We call an entry of the tuple space an **Entry**[^tuple] and call **View** a set of Entry.
+
+[^tuple]: We use Entry instead of Tuple not to conflict with the use of tuple in Quint.
+
+Let $v_1, v_2, v_3$ be **View**.
+We say that $v_1$ is superseded by $v_2$ if all elements of $v_1$ are rendered stale by some subset of $v_2$ and note this relation as $v_1 < v_2$.
```

Comment:
again, the definition of stale is missinig
imho, this should be something like:
consider two entries e=(h,r,s,v,p) and e'=(h',r',s',v',p')
then
e < e'= \lor h < h' // still unclear to me if we need h as a parameter
    	\lor h=h' \land r<r' \land PoER(r) // PoER is a proof that round r has ended
	\lor h=h' \land r=r' \land s<s' // with the usual order PROPOSAL < PRE-VOTE < PRE-COMMIT
by extension < can also be defined naturally across views
in which case, v < v' = \forall e \in v. \exists e' \in v'. e < e'